File: src/chapters/nonlinear-equations/content/code/fixed-point.cpp
#include <iostream>
#include <functional>
#include <cmath>
using namespace std;

// Fixed-point iteration x_{k+1} = g(x_k).
double fixed_point(function<double(double)> g, double x0, double tol = 1e-12, int max_iter = 200) {
    double x = x0;
    for (int k = 0; k < max_iter; ++k) {
        double xn = g(x);
        if (fabs(xn - x) < tol) return xn;
        x = xn;
    }
    return x;
}

int main() {
    cout.precision(12);
    cout << fixed_point([](double x) { return cos(x); }, 1.0) << "\n";
}

File: src/pages/practice/guides/matlab/fixed-point.m
function [x, k] = fixed_point(g, x0, tol, max_iter)
% FIXED_POINT  Iteration x_{k+1} = g(x_k).
    if nargin < 3, tol = 1e-12; end
    if nargin < 4, max_iter = 200; end
    x = x0;
    for k = 1:max_iter
        xn = g(x);
        if abs(xn - x) < tol, x = xn; return; end
        x = xn;
    end
end

% --- Demo ---
disp(fixed_point(@cos, 1.0));               % ~0.739085
